express/parsers/apps/aiida/formats/zip.py

```diff
@@ -0,0 +1,171 @@
+    """
+    Parser for AiiDA archive zip files.
+
+    Args:
+        zip_file_path: path to AiiDA archive file
+
+    Raises:
+        BadZipFile:
+            The file at the given path cannot be read as a valid zip file.
+        BadAiidaArchiveZipFile:
+            The file at the give path cannot be read as valid AiiDA archive file.
+    """
+
+    def __init__(self, zip_file_path):
+        self.zip_file_path = zip_file_path
+
+        self._metadata = None
+        self._data = None
+
+        self._check_supported_versions()  # triggers parsing of metadata
+
+    @property
+    def metadata(self):
+        if self._metadata is None:
+            with ZipFile(self.zip_file_path) as source:
+                self._metadata = json.loads(source.read('metadata.json'))
+        return self._metadata
+
+    @property
+    def data(self):
+        if self._data is None:
+            with ZipFile(self.zip_file_path) as source:
+                self._data = json.loads(source.read('data.json'))
+        return self._data
+
+    def _check_supported_versions(self):
+        """
+        Check whether the versions of the exported archive file are supported.
+
+        Args:
+            metadata (dict): deserialized metadata.json object
+
+        Raises:
+            RuntimeError
+        """
+        try:
+            aiida_version = self.metadata['aiida_version']
+            export_version = self.metadata['export_version']
+        except KeyError as error:
+            raise BadAiidaArchiveZipFile(self.zip_file_path, error)
+
+        if self.metadata['aiida_version'] != SUPPORTED_AIIDA_VERSION:
+            raise AiidaArchiveFileParseError(
+                path=self.zip_file_path,
+                reason=f"aiida_version ({metadata['aiida_version']}) not supported. "
+                f"Supported: {SUPPORTED_AIIDA_VERSION}")
+
+        if self.metadata['export_version'] != SUPPORTED_AIIDA_ARCHIVE_VERSION:
+            raise AiidaArchiveFileParseError(
+                path=self.zip_file_path,
+                rason=f"archive export_version ({metadata['export_version']}) not supported. "
+                f"Supported: {SUPPORTED_AIIDA_ARCHIVE_VERSION}")
+
+    def structures(self):
+        """
+        Extract all structures from compatible AiiDA archive zip files.
+
+        Returns:
+            list
+        """
+
+        return list(self._gather_structures(self.data))
+
+    @classmethod
+    def _gather_structures(cls, data):
+        """
+        Yield parsed structure objects from nodes within AiiDA archive.
+
+        Args:
+            data (dict): deserialized data.json object
+        Yields:
+            dict
+        """
+
+        nodes = data['export_data']['Node']
+        structure_nodes = {pk: node for (pk, node) in nodes.items()
+                           if node['node_type'] == 'data.structure.StructureData.'}
+
+        for pk in structure_nodes:
+            export_data = data['export_data']['Node'][pk]
```

Comment:
why not just `for pk, export_data in structure_nodes.items():`  Doesn't that do the same thing?

Reply:
Agreed, I refactored that a little bit.